osh/glob_.py

```diff
@@ -96,6 +91,23 @@ def GlobEscape(s):
     return pyutil.BackslashEscape(s, GLOB_META_CHARS)
 
 
+def GlobEscapeUnquotedSubstitution(s):
+    # type: (str) -> str
+    """For the substitution results such as $v with v='a\*b.txt'."""
+
+    # When glob expansion is not performed, unquoted substitutions need to be
+    # literally preserved in the final result.  A backslash inside unquoted
+    # substitutions should be treated as escaping in glob expansion, yet it
+    # needs to be preserved in the final result without glob expansion.  This
+    # means that a backslash inside unquoted substitutions is treated
+    # differently from quoted backslash, so we need a distinct representaton.
+    # We choose \@ here.
```

Comment:
> the data structures are wrong

Yeah, that was exactly the case. The issue is explained in this code comment. In short, the two backslashes in `'a\b'$v` (with `v='c\d'`) have different behaviors, but they are both escaped as `\\` in the old representation.

Reply:
OK well I guess I'm saying I would like to change the data structures even more, so the code is simpler, but we might not know what that will be yet

This area is hard, because many things depend on each other!